File: src/test/java/cube8540/oauth/authentication/credentials/oauth/security/endpoint/DefaultRedirectResolverTest.java
package cube8540.oauth.authentication.credentials.oauth.security.endpoint;

import cube8540.oauth.authentication.credentials.oauth.error.InvalidRequestException;
import cube8540.oauth.authentication.credentials.oauth.error.RedirectMismatchException;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.oauth2.core.OAuth2Error;
import org.springframework.security.oauth2.core.OAuth2ErrorCodes;

import java.net.URI;
import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@DisplayName("기본 리다이렉트 주소 Resolver 클래스 테스트")
class DefaultRedirectResolverTest {

    private DefaultRedirectResolver resolver;

    @BeforeEach
    void setup() {
        this.resolver = new DefaultRedirectResolver();
    }

    @Test
    @DisplayName("요청 받은 URI가 null 이며 클라이언트에 등록된 리다이렉트 주소가 한개일 때")
    void whenRequestedUriIsNullAndSingleRedirectUriRegisteredOnClient() {
        Set<URI> redirectURI = new HashSet<>();
        OAuth2ClientDetails clientDetails = mock(OAuth2ClientDetails.class);

        redirectURI.add(URI.create("http://localhost:8080"));
        when(clientDetails.getRegisteredRedirectUris()).thenReturn(redirectURI);

        URI result = resolver.resolveRedirectURI(null, clientDetails);
        assertEquals(URI.create("http://localhost:8080"), result);
    }

    @Test
    @DisplayName("요청 받은 URI가 null 이며 클라이언트에 등록된 리다이렉트 주소가 두개 이상일 때")
    void whenRequestedUriIsNullAndMoreThanOneRegisteredUriOnClient() {
        Set<URI> redirectURI = new HashSet<>();
        OAuth2ClientDetails clientDetails = mock(OAuth2ClientDetails.class);

        redirectURI.add(URI.create("http://localhost:8080"));
        redirectURI.add(URI.create("http://localhost:8081"));
        redirectURI.add(URI.create("http://localhost:8082"));
        when(clientDetails.getRegisteredRedirectUris()).thenReturn(redirectURI);

        OAuth2Error error = assertThrows(InvalidRequestException.class, () -> resolver.resolveRedirectURI(null, clientDetails)).getError();
        assertEquals(OAuth2ErrorCodes.INVALID_REQUEST, error.getErrorCode());
    }

    @Test
    @DisplayName("요청 받은 URI가 등록 되지 않은 URI 일 때")
    void whenRequestedUriIsNotRegisteredInClient() {
        Set<URI> redirectURI = new HashSet<>();
        OAuth2ClientDetails clientDetails = mock(OAuth2ClientDetails.class);

        redirectURI.add(URI.create("http://localhost:8080"));
        redirectURI.add(URI.create("http://localhost:8081"));
        redirectURI.add(URI.create("http://localhost:8082"));
        when(clientDetails.getRegisteredRedirectUris()).thenReturn(redirectURI);

        String uri = "http://localhost:8085";
        assertThrows(RedirectMismatchException.class, () -> resolver.resolveRedirectURI(uri, clientDetails));
    }

    @Test
    @DisplayName("요청 받은 URI가 등록 되어 있는 URI 일 떄")
    void whenRequestedUriIsRegisteredInClient() {
        Set<URI> redirectURI = new HashSet<>();
        OAuth2ClientDetails clientDetails = mock(OAuth2ClientDetails.class);

        redirectURI.add(URI.create("http://localhost:8080"));
        redirectURI.add(URI.create("http://localhost:8081"));
        redirectURI.add(URI.create("http://localhost:8082"));
        when(clientDetails.getRegisteredRedirectUris()).thenReturn(redirectURI);

        String uri = "http://localhost:8080";
        URI result = resolver.resolveRedirectURI(uri, clientDetails);
        assertEquals(URI.create(uri), result);
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/security/endpoint/DefaultRedirectResolverTest.kt
package cube8540.oauth.authentication.oauth.security.endpoint

import cube8540.oauth.authentication.oauth.error.InvalidRequestException
import cube8540.oauth.authentication.oauth.error.RedirectMismatchException
import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Test
import org.springframework.security.oauth2.core.OAuth2ErrorCodes
import java.net.URI

class DefaultRedirectResolverTest {

    private val resolver = DefaultRedirectResolver()

    @Test
    fun `requested redirect uri is null and client has one redirect uri`() {
        val clientDetails: OAuth2ClientDetails = mockk {
            every { registeredRedirectUris } returns setOf(URI.create("http://localhost/callback"))
        }

        val result = resolver.resolveRedirectURI(null, clientDetails)
        assertThat(result).isEqualTo(URI.create("http://localhost/callback"))
    }

    @Test
    fun `requested redirect uri is null and client has several redirect uri`() {
        val registeredRedirectUriSet = setOf(URI.create("http://localhost/callback1"),
            URI.create("http://localhost/callback2"),
            URI.create("http://localhost/callback3"))
        val clientDetails: OAuth2ClientDetails = mockk {
            every { registeredRedirectUris } returns registeredRedirectUriSet
        }

        val thrown = catchThrowable { resolver.resolveRedirectURI(null,  clientDetails) }
        assertThat(thrown).isInstanceOf(InvalidRequestException::class.java)
        assertThat((thrown as InvalidRequestException).error.errorCode).isEqualTo(OAuth2ErrorCodes.INVALID_REQUEST)
    }

    @Test
    fun `requested redirect uri is not registered in client`() {
        val registeredRedirectUriSet = setOf(URI.create("http://localhost/callback1"),
            URI.create("http://localhost/callback2"),
            URI.create("http://localhost/callback3"))
        val clientDetails: OAuth2ClientDetails = mockk {
            every { registeredRedirectUris } returns registeredRedirectUriSet
        }

        val thrown = catchThrowable { resolver.resolveRedirectURI("http://localhost/different", clientDetails) }
        assertThat(thrown).isInstanceOf(RedirectMismatchException::class.java)
    }

    @Test
    fun `requested redirect uri is registered in client`() {
        val registeredRedirectUriSet = setOf(URI.create("http://localhost/callback1"),
            URI.create("http://localhost/callback2"),
            URI.create("http://localhost/callback3"))
        val clientDetails: OAuth2ClientDetails = mockk {
            every { registeredRedirectUris } returns registeredRedirectUriSet
        }

        val result = resolver.resolveRedirectURI("http://localhost/callback1", clientDetails)
        assertThat(result).isEqualTo(URI.create("http://localhost/callback1"))
    }
}